model/jpa/src/main/java/org/keycloak/models/jpa/RoleAdapter.java

```diff
@@ -89,30 +94,74 @@ public void setName(String name) {
 
     @Override
     public boolean isComposite() {
-        return getCompositesStream().count() > 0;
+        TypedQuery<String> query = em.createNamedQuery("getChildrenRoleIds", String.class);
+        query.setParameter("roleId", getId());
+        query.setMaxResults(1);
+        return !query.getResultList().isEmpty();
     }
 
     @Override
     public void addCompositeRole(RoleModel role) {
+        // Avoid lazy loading the composite role collection if not already done
+        if (Persistence.getPersistenceUtil().isLoaded(getEntity(), "compositeRoles")) {
```

Comment:
Could you please check a similar change in this PR https://github.com/keycloak/keycloak/pull/10623?

Basically, using `PersistenceUtil` is not reliable, and the better is use `EntityManager.contains` instead.

Reply:
HI @pedroigor ,
Thanks for the reference, I learned something (or did not remember that it existed in the first place). :-)

What's weird is that this JPA method is recommended in Hibernate documentation https://docs.jboss.org/hibernate/orm/5.6/userguide/html_single/Hibernate_User_Guide.html#pc-contains :

> In JPA there is an alternative means to check laziness using the following javax.persistence.PersistenceUtil pattern (which is recommended wherever possible).

Method `EntityManager.contains` cannot be used here because this method applies to entities, not internal fields holding relationship references.

However, the method `Hibernate.isPropertyInitialized(Object, String)` can be used successfully as a replacement.

Do you think it would be better?

Reply:
I found that problem with `PersistenceUtil` when running the server using Quarkus. If you can make sure the method does what you expect when using the new dist, we should be fine.

Reply:
Well, I did not see any issue (what was your symptom?) but the Hibernate code itself hints that there could be a problem with `PersistenceUtil.isLoaded(Object, String)` (the ref mentioned in [#10623](https://github.com/keycloak/keycloak/pull/10623)):
```
	public boolean isLoaded(Object entity, String attributeName) {
		// added log message to help with HHH-7454, if state == LoadState,NOT_LOADED, returning true or false is not accurate.
		log.debug( "PersistenceUnitUtil#isLoaded is not always accurate; consider using EntityManager#contains instead" );
```

So I'm not very comfortable to leave it as it is, I'd rather use `Hibernate.isPropertyInitialized(Object, String)` instead, as the code is straightforward and directly taps into the internal lazy loaded structure flag: https://github.com/hibernate/hibernate-orm/blob/main/hibernate-core/src/main/java/org/hibernate/Hibernate.java#L152

Ok with that?

Reply:
@pedroigor I went that route (debugging), and it turns out that you are totally correct: the `PersistenceUtil` approach still returns `false` even after the collection was loaded, which is wrong :-( .
Changed to `Hibernate.isPropertyInitialized` and verified it's now working as expected.
PR updated.